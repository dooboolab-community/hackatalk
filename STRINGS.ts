import * as Localization from 'expo-localization';

import i18n from 'i18n-js';

const en = {
  LOADING: 'Loading...',
  HELLO: 'Hello',
  LOGIN: 'Login',
  EMAIL: 'Email',
  PASSWORD: 'Password',
  SIGN_UP: 'Sign Up',
  LOGOUT: 'Logout',
  FORGOT_PW: 'Forgot password?',
  FIND_PW: 'Find Password',
  NAME: 'name',
  STATUS_MSG: 'Status message',
  REGISTER: 'Register',
  SEND_LINK: 'Send Link',
  FRIEND: 'Friend',
  MESSAGE: 'Message',
  MY_PROFILE: 'My profile',
  UPDATE: 'Update',
  NO_CONTENT: 'No content',
  ADD_FRIEND: 'Add',
  DELETE_FRIEND: 'Delete',
  SEARCH_USER: 'Search User',
  CHAT: 'Chat',
  SEND: 'Send',
  WRITE_MESSAGE: 'Write message...',
  ERROR: 'Error',
  FRIEND_ADDED: 'Added to your friend.',
  FRIEND_ALREADY_ADDED: 'Already your friend.',
  SETTING: 'Setting',
  SIGN_IN_WITH_GOOGLE: 'Sign in with Google',
  SIGN_IN_WITH_FACEBOOK: 'Sign in with Facebook',
};

const ko = {
  LOADING: '로딩중...',
  HELLO: '안녕하세요',
  LOGIN: '로그인',
  LOGOUT: '로그아웃',
  EMAIL: '이메일',
  PASSWORD: '패스워드',
  SIGN_UP: '회원가입',
  FORGOT_PW: '비밀번호를 잊어버리셨나요?',
  FIND_PW: '비밀번호 찾기',
  NAME: '이름',
  STATUS_MSG: '상태메세지',
  REGISTER: '가입',
  SEND_LINK: '링크 보내기',
  FRIEND: '친구',
  MESSAGE: '메세지',
  MY_PROFILE: '나의 프로필',
  UPDATE: '수정',
  NO_CONTENT: '컨텐츠가 없습니다',
  ADD_FRIEND: '친구추가',
  DELETE_FRIEND: '친구삭제',
  SEARCH_USER: '사용자 검색',
  CHAT: '채팅',
  SEND: '보내기',
  WRITE_MESSAGE: '메세지를 입력해주세요…',
  ERROR: '에러',
  FRIEND_ADDED: '친구로 등록되었습니다.',
  FRIEND_ALREADY_ADDED: '이미 친구입니다.',
  SETTING: '설정',
  SIGN_IN_WITH_GOOGLE: '구글 계정으로 시작하기',
  SIGN_IN_WITH_FACEBOOK: '페이스북 계정으로 시작하기',
};

i18n.fallbacks = true;
i18n.translations = { en, ko };
i18n.locale = Localization.locale;

export const getString = (param: string, mapObj?: object): string => {
  if (mapObj) {
    i18n.t(param, mapObj);
  }
  return i18n.t(param);
};
