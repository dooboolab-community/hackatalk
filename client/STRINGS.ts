import * as Localization from 'expo-localization';

import en from './assets/langs/en.json';
import i18n from 'i18n-js';
import ko from './assets/langs/ko.json';

i18n.fallbacks = true;
i18n.translations = { en, ko };
i18n.locale = Localization.locale;

export const getString = (param: string, mapObj?: Record<string, unknown>): string => {
  if (mapObj) {
    return i18n.t(param, mapObj);
  }

  return i18n.t(param);
};
