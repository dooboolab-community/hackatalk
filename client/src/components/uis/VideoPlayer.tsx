import * as VideoThumbnails from 'expo-video-thumbnails';

import {AVPlaybackStatus, Video} from 'expo-av';
import {Image, Platform, View} from 'react-native';
import React, {FC, useEffect, useRef, useState} from 'react';

import {LoadingIndicator} from 'dooboo-ui';
import {TouchableOpacity} from 'react-native-gesture-handler';
import {getString} from '../../../STRINGS';
import styled from '@emotion/native';

type Props = {
  uri: string;
  setMediaError: (error: string) => void;
};

const StyledVideoLoadText = styled.Text`
  font-size: 14px;
  font-weight: bold;
  color: ${({theme}) => theme.focused};
  background-color: 'rgba(255,255,255, 0.5)';
  padding: 2px 5px 2px 5px;
`;

const StyledLoadErrorText = styled.Text`
  font-size: 12px;
  color: ${({theme}) => theme.errorTitle};
  align-self: center;
`;

const StyledThumbnailWrapper = styled.View`
  position: absolute;
  top: 0;
  left: 0;
  width: 240px;
  height: 150px;
  justify-content: center;
  align-items: center;
  z-index: 100;
`;

const VideoPlayer: FC<Props> = ({uri, setMediaError}) => {
  const videoRef = useRef<Video | null>(null);

  const [loadStarted, setLoadStarted] = useState(false);

  const [status, setStatus] = useState<AVPlaybackStatus>({
    isLoaded: false,
  });

  const [thumbnail, setThumbnail] =
    useState<VideoThumbnails.VideoThumbnailsResult | null>(null);

  const [thumbnailError, setThumbnailError] = useState('');

  useEffect(() => {
    let video = videoRef.current;

    if (Platform.OS !== 'web')
      (async () => {
        try {
          const result = await VideoThumbnails.getThumbnailAsync(uri, {
            time: 0,
            quality: 0.5,
          });
          setThumbnail(result);
        } catch (e) {
          setThumbnailError(getString('THUMBNAIL_ERROR'));
        }
      })();

    return () => {
      /* istanbul ignore else */
      if (video) video.unloadAsync();
    };
  }, [uri]);

  const handleLoadPress = async (): Promise<void> => {
    setLoadStarted(true);

    await videoRef.current?.loadAsync(
      {uri},
      {
        progressUpdateIntervalMillis: 500,
        positionMillis: 0,
        shouldPlay: true,
        rate: 1.0,
        shouldCorrectPitch: false,
        volume: 1.0,
        isMuted: false,
        isLooping: false,
      },
    );
  };

  return (
    <View testID="Video-Container">
      <Video
        ref={videoRef}
        style={{
          width: 240,
          height: 150,
        }}
        useNativeControls
        resizeMode="contain"
        onError={() => {
          setMediaError(getString('FAILED_FETCH', {media: getString('VIDEO')}));
        }}
        onPlaybackStatusUpdate={(update: AVPlaybackStatus) => {
          setStatus(update);
        }}
        {...(Platform.OS === 'web' && {source: {uri}})}
      />
      {Platform.OS !== 'web' &&
        (!loadStarted ? (
          <StyledThumbnailWrapper>
            <TouchableOpacity
              onPress={handleLoadPress}
              style={{justifyContent: 'center', alignItems: 'center', flex: 1}}
            >
              {!!thumbnail && (
                <Image
                  source={{uri: thumbnail.uri}}
                  style={{
                    width: 240,
                    height: 150,
                    position: 'absolute',
                    opacity: 0.7,
                  }}
                  resizeMode={'contain'}
                />
              )}
              <StyledVideoLoadText>
                {getString('MEDIA_LOAD', {media: getString('VIDEO')})}
              </StyledVideoLoadText>
              {!!thumbnailError && (
                <StyledLoadErrorText>{thumbnailError}</StyledLoadErrorText>
              )}
            </TouchableOpacity>
          </StyledThumbnailWrapper>
        ) : !status.isLoaded ? (
          <LoadingIndicator />
        ) : null)}
    </View>
  );
};

export default VideoPlayer;
