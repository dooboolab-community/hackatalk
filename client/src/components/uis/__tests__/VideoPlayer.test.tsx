import 'react-native';

import * as React from 'react';

import {fireEvent, render, waitFor} from '@testing-library/react-native';

import VideoPlayer from '../VideoPlayer';
import {getString} from '../../../../STRINGS';

const props = {
  uri: 'http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/ElephantsDream.mp4',
  setMediaError: jest.fn(),
};

describe('[VideoPlayer] render', () => {
  it('render without crashing', async () => {
    const screen = render(<VideoPlayer {...props} />);
    const json = screen.toJSON();

    await waitFor(() => {
      expect(json).toBeTruthy();
      expect(json).toMatchSnapshot();
    });
  });

  it('should display thumbnail error when uri was not provided', async () => {
    const newProps = {...props, uri: undefined};
    const {getByText} = render(<VideoPlayer {...newProps} />);

    await waitFor(() => {
      expect(getByText(getString('THUMBNAIL_ERROR'))).toBeTruthy();
    });
  });

  it('should call onPlaybackStatusUpdate when loadAsync Resolved', async () => {
    const screen = render(<VideoPlayer {...props} />);

    await waitFor(async () => {
      const button = screen.getByText(
        getString('MEDIA_LOAD', {media: getString('VIDEO')}),
      );
      fireEvent.press(button);
    });

    expect(screen.toJSON().children[0]).toBe('Video isLoaded? : true');
  });

  it('should call onError when loadAsync rejected', async () => {
    const newProps = {...props, uri: undefined};

    const {getByText} = render(<VideoPlayer {...newProps} />);

    await waitFor(() => {
      const button = getByText(
        getString('MEDIA_LOAD', {media: getString('VIDEO')}),
      );
      fireEvent.press(button);
    });

    expect(props.setMediaError).toBeCalled();
  });
});

describe('[VideoPlayer] web render', () => {
  it('render without crashing', async () => {
    jest.doMock('react-native/Libraries/Utilities/Platform', () => {
      return {OS: 'web'};
    });

    const {queryByText} = render(<VideoPlayer {...props} />);

    expect(
      queryByText(getString('MEDIA_LOAD', {media: getString('VIDEO')})),
    ).toBeNull();
  });
});
