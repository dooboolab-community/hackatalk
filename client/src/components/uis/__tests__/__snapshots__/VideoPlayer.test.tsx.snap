// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[VideoPlayer] render render without crashing 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View>
    Video isLoaded? : false
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "height": 150,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "top": 0,
            "width": 240,
            "zIndex": 100,
          },
          undefined,
        ]
      }
    >
      <RNGestureHandlerButton
        collapsable={false}
        exclusive={true}
        handlerTag={1}
        handlerType="NativeViewGestureHandler"
        onGestureEvent={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onHandlerStateChange={[Function]}
        rippleColor={0}
        touchSoundDisabled={false}
      >
        <View
          accessible={true}
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "backgroundColor": "'rgba(255,255,255, 0.5)'",
                  "color": "#fff",
                  "fontSize": 14,
                  "fontWeight": "bold",
                  "paddingBottom": 2,
                  "paddingLeft": 5,
                  "paddingRight": 5,
                  "paddingTop": 2,
                },
                undefined,
              ]
            }
          >
            Touch to Load Video
          </Text>
        </View>
      </RNGestureHandlerButton>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
